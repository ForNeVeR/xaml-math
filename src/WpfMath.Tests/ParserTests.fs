module WpfMath.Tests.ParserTests

open Xunit

open WpfMath.Tests.ApprovalTestUtils

[<Fact>]
let ``2+2``() =
    verifyParseResult "2+2"

[<Theory>]
[<InlineData("(", ")", "(", ")")>]
[<InlineData("[", "]", "lbrack", "rbrack")>]
[<InlineData("{", "}", "lbrace", "rbrace")>]
[<InlineData("<", ">", "langle", "rangle")>]
let delimiters(left : string, right : string, lResult : string, rResult : string) =
    verifyParseResultScenario
    <| sprintf "%s,%s" lResult rResult
    <| sprintf @"\left%sa\right%s" left right

[<Theory>]
[<InlineData(".", ")", true, false)>]
[<InlineData("(", ".", false, true)>]
let emptyDelimiters(left : string, right : string, isLeftEmpty : bool, isRightEmpty : bool) =
    verifyParseResultScenario
    <| sprintf "(%s,%s,%A,%A)" left right isLeftEmpty isRightEmpty
    <| sprintf @"\left%sa\right%s" left right

[<Fact>]
let unmatchedDelimiters() =
    verifyParseResult @"\left)a\right|"

[<Fact>]
let expressionInBraces() =
    verifyParseResult @"\left(2+2\right)"

[<Fact>]
let expressionAfterBraces() =
    verifyParseResult @"\left(2+2\right) + 1"

[<Fact>]
let textCommand() =
    verifyParseResult @"\text{test}"

[<Fact>]
let spacesInText() =
    verifyParseResult @"\text{a b c}"

[<Fact>]
let cyrillicText() =
    verifyParseResult @"\text{абв}"

[<Fact>]
let underscoreText() =
    verifyParseResult @"\text{_}"

[<Fact>]
let commandsInText() =
    verifyParseResult @"\text{\alpha \beta \unknowncommand}"

[<Fact>]
let mathrm() =
    verifyParseResult @"\mathrm{sin}"

[<Fact>]
let complexMathrm() =
    verifyParseResult @"\mathrm{\left(2+2\right)} + 1"

[<Fact>]
let mathit() =
    verifyParseResult @"\mathit{sin}"

[<Fact>]
let mathcal() =
    verifyParseResult @"\mathcal{sin}"

[<Fact>]
let lim() =
    verifyParseResult @"\lim_{n} x"

[<Fact>]
let limInCurlyBraces() =
    verifyParseResult @"{\lim} x"

[<Fact>]
let sin() =
    verifyParseResult @"\sin^{n} x"

[<Fact>]
let intF() =
    verifyParseResult @"\int f"

[<Fact>]
let emptyCurlyBraces() =
    verifyParseResult @"{}"

[<Fact>]
let delimiterWithScripts() =
    verifyParseResult @"\left(2+2\right)_a^b"

[<Fact>]
let textWithExpression() =
    verifyParseResult @"\text{2+2}"

[<Theory>]
[<InlineData("{red}1123");
  InlineData("{red}{1}123");
  InlineData("{red} 1123");
  InlineData("{red} {1}123")>]
let color(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\color%s" text

[<Theory>]
[<InlineData(@"{red}1123");
  InlineData(@"{red}{1}123");
  InlineData(@"{red} 1123");
  InlineData(@"{red} {1}123")>]
let colorbox(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\colorbox%s" text

[<Theory>]
[<InlineData("2x123");
  InlineData("2{x}123");
  InlineData("{2}x123");
  InlineData("{2}{x}123");
  InlineData(" 2 x123");
  InlineData("2 {x}123");
  InlineData(" 2{x}123")>]
let frac(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\frac%s" text

[<Theory>]
[<InlineData("1123");
  InlineData("{1}123");
  InlineData(" 1123");
  InlineData(" {1}123")>]
let overline(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\overline%s" text

[<Theory>]
[<InlineData("1123");
  InlineData("{1}123");
  InlineData(" 1123");
  InlineData(" {1}123")>]
let sqrt(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\sqrt%s" text

[<Theory>]
[<InlineData(" [2]1123");
  InlineData(" [ 2]{1}123");
  InlineData("[2 ] 1123");
  InlineData("[ 2 ] {1}123")>]
let sqrtWithOptArgument(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\sqrt%s" text

[<Theory>]
[<InlineData("1123");
  InlineData("{1}123");
  InlineData(" 1123");
  InlineData(" {1}123")>]
let underline(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\underline%s" text

[<Theory>]
[<InlineData("x^y_z");
  InlineData("x^y_{z}");
  InlineData("x^{y}_z");
  InlineData("x^{y}_{z}");
  InlineData("x^y_ z");
  InlineData("x ^ {y} _ {z}")>]
let scripts(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| text

[<Theory>]
[<InlineData(" 1123");
  InlineData(" {1}123")>]
let textArgumentParsing(text : string) : unit =
    verifyParseResultScenario
    <| processSpecialChars text
    <| sprintf @"\text%s" text

[<Fact>]
let hat() : unit =
    verifyParseResult @"{\hat T}"

[<Fact>]
let integral() =
    verifyParseResult @"\int_a^b"
