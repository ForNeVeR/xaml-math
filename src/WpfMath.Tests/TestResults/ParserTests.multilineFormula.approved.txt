{
  "TextStyle": null,
  "RootAtom": {
    "MatrixCells": [
      [
        {
          "PreviousAtom": null,
          "Elements": [
            {
              "Character": "l",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 0,
                "End": 1,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "i",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 1,
                "End": 2,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "n",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 2,
                "End": 3,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "e",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 3,
                "End": 4,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "1",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 5,
                "End": 6,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            }
          ],
          "Type": "Ordinary",
          "Source": {
            "Start": 0,
            "End": 6,
            "Length": 6,
            "Source": "line 1\\\\line 2\\\\line 3",
            "SourceName": "User input"
          }
        }
      ],
      [
        {
          "PreviousAtom": null,
          "Elements": [
            {
              "Character": "l",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 8,
                "End": 9,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "i",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 9,
                "End": 10,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "n",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 10,
                "End": 11,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "e",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 11,
                "End": 12,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "2",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 13,
                "End": 14,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            }
          ],
          "Type": "Ordinary",
          "Source": {
            "Start": 8,
            "End": 14,
            "Length": 6,
            "Source": "line 1\\\\line 2\\\\line 3",
            "SourceName": "User input"
          }
        }
      ],
      [
        {
          "PreviousAtom": null,
          "Elements": [
            {
              "Character": "l",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 16,
                "End": 17,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "i",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 17,
                "End": 18,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "n",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 18,
                "End": 19,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "e",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 19,
                "End": 20,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            },
            {
              "Character": "3",
              "TextStyle": null,
              "IsDefaultTextStyle": true,
              "IsTextSymbol": false,
              "Type": "Ordinary",
              "Source": {
                "Start": 21,
                "End": 22,
                "Length": 1,
                "Source": "line 1\\\\line 2\\\\line 3",
                "SourceName": "User input"
              }
            }
          ],
          "Type": "Ordinary",
          "Source": {
            "Start": 16,
            "End": 22,
            "Length": 6,
            "Source": "line 1\\\\line 2\\\\line 3",
            "SourceName": "User input"
          }
        }
      ]
    ],
    "VerticalPadding": 0.35,
    "HorizontalPadding": 0.35,
    "MatrixCellAlignment": "Left",
    "Type": "Ordinary",
    "Source": null
  },
  "Source": {
    "Start": 0,
    "End": 22,
    "Length": 22,
    "Source": "line 1\\\\line 2\\\\line 3",
    "SourceName": "User input"
  }
}